### ISSUE TITLE
Deprecation and removal of the `SecurityContextDeny` admission plugin

### ISSUE THREAD

**Issue by mtardy** [kubernetes/kubernetes#111516]
**tl;dr** the `SecurityContextDeny` admission plugin is going to be removed because it's outdated and now replaced by the `PodSecurity` admission.

#### Context

While digging on PodSecurityPolicy history, I came across this 2015 PR https://github.com/kubernetes/kubernetes/pull/7343 that added the `SecurityContext` field in containers spec. It also added a new admission plugin called **scdeny**. In fact, this PR was remade after ServiceAccounts were added, it was first proposed here https://github.com/kubernetes/kubernetes/pull/6287 and a comment asked for adding the admission control part (with a deny all and an allow all, for testing purposes), which should be enabled by default https://github.com/kubernetes/kubernetes/pull/6287#issuecomment-95715700. 

In the end, when https://github.com/kubernetes/kubernetes/pull/7343 was merged, only scdeny was added and it was disabled by default. At this time `SecurityContext` was only composed of the pre-existing `Capabilities` and `Privileged` fields and the newly introduced `RunAsUser` and `SELinuxOptions`. scdeny blocked the use of the new `RunAsUser` and `SELinuxOptions` fields. There was an issue opened afterward on enabling scdeny by default https://github.com/kubernetes/kubernetes/issues/16456.

This admission control plugin survived until now, [is still available]( https://github.com/kubernetes/kubernetes/blob/15d12e713f82672ac8951ed14cb5e83ad3288d8b/plugin/pkg/admission/securitycontext/scdeny/admission.go) and was, for example, updated in 2019. Its code is super small and its [validate function](https://github.com/kubernetes/kubernetes/blob/15d12e713f82672ac8951ed14cb5e83ad3288d8b/plugin/pkg/admission/securitycontext/scdeny/admission.go#L54) is short enough to be easy to understand by just reading it.

With the new ephemeral containers subresource, this plugin misses its security context. I asked about that topic in a [Slack thread](https://kubernetes.slack.com/archives/C019LFTGNQ3/p1658755949311999) in sig-security channel, we discussed that topic at sig-security meeting and it led to the creation of this issue.

I guess with this situation we have two solutions:
- We update it, using the utility to iterate over all the containers (as suggested [here](https://kubernetes.slack.com/archives/C019LFTGNQ3/p1659023107113069?thread_ts=1658755949.311999&cid=C019LFTGNQ3)) making it ephemeral containers "aware".
- We deprecate it and we remove it because it's a bad, not configurable, specific admission control plugin for today's Kubernetes that is not really relevant anymore. At the time, [its design proposal](https://github.com/kubernetes/design-proposals-archive/blob/main/auth/security_context.md) explained its reason for existence:
> It is up to an admission plugin to determine if the security context is acceptable or not. At the time of writing, the admission control plugin for security contexts will only allow a context that has defined capabilities or privileged. Contexts that attempt to define a UID or SELinux options will be denied by default. In the future the admission plugin will base this decision upon configurable policies that reside within the [service account](http://pr.k8s.io/2297).

Most of the people I talked with are in favor of the second option (sig security and some sig auth members) but it would be nice to gather more thoughts. If we decide to remove it, then it would be a bit particular because this thing was added before KEPs were a thing in Kubernetes. It was still "design proposals" which are archived, and the one on the security context is [available here](https://github.com/kubernetes/design-proposals-archive/blob/main/auth/security_context.md). So let's discuss on this issue and if we have enough feedback, we could create a KEP associated with this, and I would be glad to handle this!

#### Action items

```[tasklist]
### Action items identified in sig-auth meeting on [2023-01-04](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.xm7t8yi1p33u)
- [x] Update documentation at https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny to note shortcomings with this admission plugin. https://github.com/kubernetes/website/pull/39131
- [x] Add a warning at runtime if the admission plugin is used with a reference to the docs of shortcomings and put the admission plugin registration behind the `SecurityContextDeny` feature gate. https://github.com/kubernetes/kubernetes/pull/115879
- [x] Update doc to reflect deprecation and add the `SecurityContextDeny` feature gate to the reference. https://github.com/kubernetes/website/pull/40166
- [x] If no usage is found or remains after reaching out to users, consider announcing it as deprecated and removing it after the deprecation period
- [x] Remove it from k/test-infra: https://github.com/kubernetes/test-infra/pull/29361
- [x] Remove it from k/k: https://github.com/kubernetes/kubernetes/pull/122612
- [x] Document the removal for v1.30: https://github.com/kubernetes/website/pull/45136
```

**Comment by sftim** [comment]
(We shouldn't ship ineffective security controls)

**Comment by tallclair** [comment]
I'm +1 for removal. I think the Pod Security Admission restricted profile captures what this plugin was trying to achieve in a much better and up-to-date way. I can't think of any reason to use this one at this point (it's very incomplete).

**Comment by sftim** [comment]
For readability
/retitle Update, or deprecation and removal, of the `scdeny` admission plugin

**Comment by sftim** [comment]
Actually, sorry
/retitle Update, or deprecation and removal, of the `SecurityContextDeny` admission plugin

**Comment by tabbysable** [comment]
I'm +1 for removal; if it was fully-functional according to its name (prohibit all pods that specify a non-default SecurityContext) it would be impossible to run a modern cluster with it (e.g. CNI daemonsets would be blocked). The fact that it is only partially functional is even more reason to remove it.

**Comment by enj** [comment]
I have added this to the Jan 2023 SIG Auth meeting agenda to discuss.  I expect we will decide to remove this in v1.27.

**Comment by mtardy** [comment]
Thanks a lot, I would love to as well, unfortunately, I could not attend the last meeting and many were canceled recently :D I will try my best to be here in January!

**Comment by liggitt** [comment]
updated the description with action items to capture shortcomings, warn on use, and sweep for public use

**Comment by mtardy** [comment]
Thanks liggitt!
I, yet again, could not join the meeting!

I was going to write a KEP under the advice of sig security people, you (or SIG auth if you talked about that) think it's not necessary and we could directly do the action items?

/assign

**Comment by liggitt** [comment]
we typically use retroactive KEPs for long-beta features we want to promote... I don't think a KEP is needed for these action items

**Comment by sftim** [comment]
@mtardy, thanks for the docs update.

**Comment by sftim** [comment]
> Add a warning at runtime if the admission plugin is used with a reference to the docs of shortcomings

To me, I think the deprecation is strong enough that we could set a warning header on any Pod create that happens when the plugin is merely active (no matter whether the Pod was admitted or denied).

**Comment by sowmyav27** [comment]
@mtardy @sftim Hi, I am part of Bug Triage Shadow team for k8s 1.27 release. I wanted to confirm this is on track for 1.27?

**Comment by mtardy** [comment]
> @mtardy @sftim Hi, I am part of Bug Triage Shadow team for k8s 1.27 release. I wanted to confirm this is on track for 1.27?

Yep it should be, I need to merge this PR https://github.com/kubernetes/kubernetes/pull/115879 for the first step to be complete for 1.27. For this version, it will be only a warning and no removal. It's 50% done basically because the first PR was merged.

**Comment by mtardy** [comment]
For the update, the code changes https://github.com/kubernetes/kubernetes/pull/115879, second step has been merged, so this is ready for v1.27. I mostly need to do a small k/website change to explicitly say that the admission plugin is deprecated and it's done for this step.

**Comment by liggitt** [comment]
work for 1.27 is complete, moving to 1.28

/milestone v1.28

**Comment by hailkomputer** [comment]
Hello @mtardy, v1.28 Bug Triage Shadow here checking in. It is still early in the release cycle, but is this still on track for the v1.28 release?

**Comment by hailkomputer** [comment]
Hello! Bug triage team shadow here. I want to check the status. The code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about three weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the issue is tagged for 1.28, is it still planned for this release?

**Comment by mtardy** [comment]
> Hello! Bug triage team shadow here. I want to check the status. The code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about three weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
> 
> As the issue is tagged for 1.28, is it still planned for this release?

👋  I'm sorry for not responding to your previous message. I would like to merge this for v1.28 but it's a bit unsure I would say. I will try :)

**Comment by mtardy** [comment]
> > Hello! Bug triage team shadow here. I want to check the status. The code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about three weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
> > As the issue is tagged for 1.28, is it still planned for this release?
> 
> 👋 I'm sorry for not responding to your previous message. I would like to merge this for v1.28 but it's a bit unsure I would say. I will try :)

Related to that, happy to say that we made some progress and finally merged https://github.com/kubernetes/website/pull/40166. Let's create the removal PRs now.

**Comment by hailkomputer** [comment]
Hi @mtardy! Bug triage team shadow here again :).

We are currently in Week 9 of the v1.28 release cycle, and the code freeze is happening next week (Week 10) at [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)

Please ensure all necessary contributions are made before the deadline if the Issue still targets v1.28.

Let me know if you need any assistance.

**Comment by mtardy** [comment]
> Hi @mtardy! Bug triage team shadow here again :).
> 
> We are currently in Week 9 of the v1.28 release cycle, and the code freeze is happening next week (Week 10) at [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)
> 
> Please ensure all necessary contributions are made before the deadline if the Issue still targets v1.28.
> 
> Let me know if you need any assistance.

reasonably this won't be ready for next Wednesday unfortunately, so let's target v1.29.

**Comment by mtardy** [comment]
> Hello @enj ! 👋🏼 Bug triage for the 1.29 release cycle is here! This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 2 weeks from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.
> 
> As the issue is tagged for 1.29, is it still planned for that release?

Hey, we'll need to target 1.30, I'm sorry I didn't take enough time to do this, nothing urgent but still it would be nice to finish that!

**Comment by dims** [comment]
/milestone v1.30

kicking the can, we are way past code freeze

**Comment by mtardy** [comment]
Well 18 months after the opening of this issue, I guess we can now say it's done and ready for v1.30.

Thanks, everyone for the help, after this code removal PR https://github.com/kubernetes/kubernetes/pull/122612 I can now say I had a _negative_ impact on k8s 🥰!
